swift_browser_ui_frontend/src/common/download.js

```diff
@@ -0,0 +1,230 @@
+import {
+  PUT,
+  GET,
+  DELETE,
+  getUploadEndpoint,
+} from "@/common/api";
+
+// Add a header to the ServiceWorker filesystem
+function addHeader(header, fname, fsize) {
+  navigator.serviceWorker.ready.then(reg => {
+    reg.active.postMessage({
+      cmd: "addHeader",
+      header: header,
+      fileName: fname,
+      fileSize: fsize,
+    });
+  });
+}
+
+export function beginDownload() {
+  navigator.serviceWorker.ready.then(reg => {
+    reg.active.postMessage({
+      cmd: "beginDownload",
+    });
+  });
+}
+
+export class DecryptedDownloadSession {
+  constructor(
+    active,
+    project = "",
+    objects = [],
+    container = "",
+    store,
+  ) {
+    this.object = "";
+    this.active = active;
+    this.project = project;
+    this.objects = objects;
+    this.container = container;
+    this.$store = store;
+    this.finished = false;
+    this.stream = undefined;
+    this.size = 0;
+    this.lastChunk = 0;
+    this.chunks = 0;
+    this.chunksLeft = 0;
+    this.currentFinished = false;
+    this.whitelistPath = `/cryptic/${this.active.id}/whitelist`;
+    this.endpoint = store.state.uploadEndpoint;
+
+    this.reader = undefined;
+    this.chunk = undefined;
+    this.readerDone = false;
+    this.offset = 0;
+    this.remainder = 0;
+    this.chunkBuffer = [];
+
+    this.controller = new AbortController();
+  }
+
+  async getFile(objectUrl) {
+    const response = await fetch(objectUrl);
+    let reader = response.body.getReader();
+    this.reader = reader;
+    let { value: chunk, done: readerDone } = await this.reader.read();
+
+    this.chunk = chunk;
+    this.readerDone = readerDone;
+
+    this.offset = 0;
+    this.remainder = 0;
+    this.chunkBuffer = [];
+  }
+
+  async getSlice() {
+    let chunk = this.chunk;
+    let readerDone = this.readerDone;
+    while (!this.readerDone) {
+      this.remainder = 65564 - this.chunkBuffer.length;
```

Comment:
What is 65564 for?

can't the chunk buffer be larger than that?

Reply:
In principle yes, if we then split it up to correct sizes afterwards. I'm not sure that would help with the performance issues caused by the poorly aligned block size, however. Fetch keeps its own buffer, so I don't think there's any benefit of increasing the size of this one.